pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,489 @@
+					Name:      req.Name,
+					Namespace: req.Namespace,
+				},
+			}, false, nil
+		}
+		return nil, false, err
+	}
+	return &svcImport, true, nil
+}
+
+func (r *mcsAPIServiceImportReconciler) getLocalService(ctx context.Context, req ctrl.Request) (*corev1.Service, error) {
+	var svc corev1.Service
+	if err := r.Client.Get(ctx, req.NamespacedName, &svc); err != nil {
+		if k8sApiErrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &svc, nil
+}
+
+func (r *mcsAPIServiceImportReconciler) doesNamespaceExist(ctx context.Context, req ctrl.Request) (bool, error) {
+	var ns corev1.Namespace
+	key := types.NamespacedName{Name: req.Namespace}
+
+	if err := r.Client.Get(ctx, key, &ns); err != nil {
+		if k8sApiErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func fromServiceToMCSAPIServiceSpec(svc *corev1.Service, svcExport *mcsapiv1alpha1.ServiceExport) *mcsapitypes.MCSAPIServiceSpec {
+	ports := make([]mcsapiv1alpha1.ServicePort, 0, len(svc.Spec.Ports))
+	for _, port := range svc.Spec.Ports {
+		ports = append(ports, mcsapiv1alpha1.ServicePort{
+			Name:        port.Name,
+			Protocol:    port.Protocol,
+			AppProtocol: port.AppProtocol,
+			Port:        port.Port,
+		})
+	}
+	mcsAPISvcType := mcsapiv1alpha1.ClusterSetIP
+	if svc.Spec.ClusterIP == corev1.ClusterIPNone {
+		mcsAPISvcType = mcsapiv1alpha1.Headless
+	}
+	mcsAPISvcSpec := &mcsapitypes.MCSAPIServiceSpec{
+		Name:                  svc.Name,
+		Namespace:             svc.Namespace,
+		Ports:                 ports,
+		Type:                  mcsAPISvcType,
+		SessionAffinity:       svc.Spec.SessionAffinity,
+		SessionAffinityConfig: svc.Spec.SessionAffinityConfig.DeepCopy(),
+	}
+	mcsAPISvcSpec.ExportCreationTimestamp = svcExport.CreationTimestamp
+	return mcsAPISvcSpec
+}
+
+func getOldestRemoteClusterService(svcExportByCluster operator.ServiceExportsByCluster) string {
+	oldestCluster := maps.Keys(svcExportByCluster)[0]
+	for cluster, currentSvc := range svcExportByCluster {
+		if currentSvc.ExportCreationTimestamp.Before(
+			&svcExportByCluster[oldestCluster].ExportCreationTimestamp) {
+			oldestCluster = cluster
+		}
+	}
+	return oldestCluster
+}
+
+type portMergeKey struct {
+	Protocol    corev1.Protocol
+	AppProtocol string
+	Port        int32
+}
+
+type portMergeValue struct {
+	mcsapiv1alpha1.ServicePort
+
+	clusterName             string
+	ExportCreationTimestamp metav1.Time
+}
+
+type mergedPorts map[portMergeKey]portMergeValue
+
+// mergePorts merge all the ports into a map while doing conflict resolution
+// with the oldest CreationTimestamp
+func mergePorts(svcExportByCluster operator.ServiceExportsByCluster) mergedPorts {
+	globalPorts := map[portMergeKey]portMergeValue{}
+	for cluster, mcsAPISvcSpec := range svcExportByCluster {
+		for _, port := range mcsAPISvcSpec.Ports {
+			appProtocol := ""
+			if port.AppProtocol != nil {
+				appProtocol = *port.AppProtocol
+			}
+			key := portMergeKey{
```

Comment:
IIUC the KEP, the key would be the name of the port (as you cannot have two ports with the same name), rather than the port, protocol, app protocol triple: 

> Should a port name be used for multiple non-identical (port, protocol, appProtocol) service ports by different constituent services, the conflict resolution policy will determine which values are used by the derived service.

Reply:
Hmm so the port name logic checking was more on `checkConflictExportPorts` so far BUT I did forget to check if we export a port that has the same name as another port, will check how I should do that :+1: . So far I was only checking if we export the same port but with a different name, like catching that essentially:
```
    - name: http
      protocol: TCP
      port: 80
      targetPort: 8080
```
```
    - name: not-http
      protocol: TCP
      port: 80
      targetPort: 8080
```

Reply:
Yeah, AFAIU the two aspects that must not be duplicated are the [name](https://github.com/kubernetes/kubernetes/blob/4b8e819355d791d96b7e9d9efe4cbafae2311c88/pkg/apis/core/validation/validation.go#L5503-L5512), and the [(port, protocol) tuple ](https://github.com/kubernetes/kubernetes/blob/4b8e819355d791d96b7e9d9efe4cbafae2311c88/pkg/apis/core/validation/validation.go#L5446-L5457)(appProtocol does not seem to be significant instead). 

Indeed, both of the two following services get rejected by k8s:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: same-port-proto
  namespace: default
spec:
  ports:
  - name: foo
    port: 80
    protocol: TCP
    appProtocol: ssh
    targetPort: 9898
  - name: bar
    port: 80
    protocol: TCP
    appProtocol: http
    targetPort: 9898
  selector:
    app: podinfo
---
apiVersion: v1
kind: Service
metadata:
  name: same-name
  namespace: default
spec:
  ports:
  - name: foo
    port: 80
    protocol: TCP
    targetPort: 9898
  - name: foo
    port: 81
    protocol: TCP
    targetPort: 9898
  selector:
    app: podinfo
``` 

```
$ Service "same-port-proto" is invalid: spec.ports[1]: Duplicate value: core.ServicePort{Name:"", Protocol:"TCP", AppProtocol:(*string)(nil), Port:80, TargetPort:intstr.IntOrString{Type:0, IntVal:0, StrVal:""}, NodePort:0}
$ Service "same-name" is invalid: spec.ports[1].name: Duplicate value: "foo"
```

Reply:
Reading this again I now realize that I skipped the part where app protocol is not significant :sweat_smile:.

Having two similar ports (similar meaning the tuple of port + protocol) with different app protocol is thus a byproduct of the check for two similar ports with the default kubernetes code.

However in our case we have to try merging the different ports from different clusters together so we have the context that the app protocol is different on a similar ports so I think we could raise the conflict message specifically towards this instead of checking for two similar ports.